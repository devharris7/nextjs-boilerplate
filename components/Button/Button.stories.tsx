import type { Story, Meta } from '@storybook/react/types-6-0';

import Button from './Button';
import type { ButtonProps } from './Button';

export default {
  title: 'Button',
  component: Button,
} as Meta;

export const Template: Story<ButtonProps> = (args) => <Button {...args}>Default</Button>;

export const Small: Story<ButtonProps> = Template.bind({});
Small.args = {
  size: 'sm',
};

export const Base: Story<ButtonProps> = Template.bind({});
Base.args = {
  size: 'md',
};

export const Large: Story<ButtonProps> = Template.bind({});
Large.args = {
  size: 'lg',
};

export const Disabled: Story<ButtonProps> = Template.bind({});
Disabled.args = {
  disabled: true,
};
