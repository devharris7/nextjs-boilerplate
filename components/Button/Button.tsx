import React, { HTMLAttributes } from 'react';
import classnames from 'classnames';

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  type?: 'button' | 'reset' | 'submit';
  children: React.ReactNode;
  size?: 'xs' | 'sm' | 'md' | 'lg';
  theme?: 'primary' | 'secondary' | 'accent' | 'info' | 'success' | 'warning' | 'error' | 'ghost';
  isLink?: boolean;
  variant?: 'outline' | 'circle' | 'square';
  loading?: boolean;
  disabled?: boolean;
}

const Button = ({
  type = 'button',
  children,
  theme = 'primary',
  isLink = false,
  variant,
  size = 'md',
  loading,
  disabled = false,
  onClick,
}: ButtonProps) => {
  const childElement = isLink ? <a>{children}</a> : children;

  return (
    <>
      <button
        type={type}
        disabled={disabled}
        className={classnames([
          'btn',
          { 'btn-disabled': disabled, loading: loading, [`btn-${variant}`]: variant },
          `btn-${theme}`,
          `btn-${size}`,
        ])}
        onClick={onClick}
      >
        {childElement}
      </button>
    </>
  );
};

export default Button;
