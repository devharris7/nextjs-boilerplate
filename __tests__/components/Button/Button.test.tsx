import { render } from "@testing-library/react";
import Button from "../../../components/Button/Button";

describe("Button", () => {
  it("renders a button with a tag", () => {
    const { container } = render(<Button isLink={true}>Test</Button>);
    expect(container).toMatchSnapshot();
  });
  it("renders a normal button", () => {
    const { container } = render(<Button isLink={false}>Test</Button>);
    expect(container).toMatchSnapshot();
  });
});
