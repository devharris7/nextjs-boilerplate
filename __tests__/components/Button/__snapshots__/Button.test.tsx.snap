// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a button with a tag 1`] = `
<div>
  <button
    class="btn btn-primary btn-md"
    type="button"
  >
    <a>
      Test
    </a>
  </button>
</div>
`;

exports[`Button renders a normal button 1`] = `
<div>
  <button
    class="btn btn-primary btn-md"
    type="button"
  >
    Test
  </button>
</div>
`;